### PR TITLE
Added input and output piping support

### PR DESCRIPTION
**Pull request by max619**
This feature allows us to use ffmpeg with bitmaps, streams and so on as input.

### CODE REVIEW

**Comment by max619** [comment]
Also a have only tested it working under Windows 10. There could be some issues with path to named pipes. If someone have an ability to test on other OS'es, i would really appreciate that.

File: FFMpegCore.Test/TasksExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace FFMpegCore.Test
+{
+    static class TasksExtensions
+    {
+        public static T WaitForResult<T>(this Task<T> task)
+        {
+            task.Wait();
```
Comment (WeihanLi) [inline FFMpegCore.Test/TasksExtensions.cs]: maybe `task.ConfigureAwait(false).GetAwait().GetResult()` is better?

Comment (max619) [reply FFMpegCore.Test/TasksExtensions.cs]: Sure

File: FFMpegCore/FFMPEG/Argument/Atoms/ForceFormatArgument.cs [hunk 2]
```diff
@@ -5,16 +5,17 @@ namespace FFMpegCore.FFMPEG.Argument
     /// <summary>
     /// Represents force format parameter
     /// </summary>
-    public class ForceFormatArgument : Argument<VideoCodec>
+    public class ForceFormatArgument : Argument<string>
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Argument/Atoms/ForceFormatArgument.cs:8]: Why this change? An enum makes it much clearer what options are available, to the user

Comment (rosenbjerg) [reply FFMpegCore/FFMPEG/Argument/Atoms/ForceFormatArgument.cs:8]: I think WeihanLi's flexible RawArgument would be a better solution than to untype existing arguments?

**Comment by rosenbjerg** [comment]
Really cool new functionality, nice work! Since it is such a big PR, I'll need time to review it properly

File: FFMpegCore/FFMPEG/Argument/ArgumentContainer.cs [hunk 3]
```diff
@@ -117,6 +116,16 @@ public bool ContainsKey(Type key)
             return _args.ContainsKey(key);
         }
 
+        public int CountExistedKeys(params Type[] types)
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Argument/ArgumentContainer.cs]: Suggestion:
Change this method to
```csharp
public bool ContainsOnlyOneOf(params Type[] types)
{
  return types.Count(t => _args.ContainsKey(t)) == 1;
}
```
To make it shorter and simpler

File: FFMpegCore/FFMPEG/Argument/Atoms/VideoCodecArgument.cs [hunk 4]
```diff
@@ -5,23 +5,25 @@ namespace FFMpegCore.FFMPEG.Argument
     /// <summary>
     /// Represents video codec parameter
     /// </summary>
-    public class VideoCodecArgument : Argument<VideoCodec>
+    public class VideoCodecArgument : Argument<string>
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Argument/Atoms/VideoCodecArgument.cs:8]: Wouldn't it be a better solution to extend the enum with new members instead?

File: FFMpegCore/FFMPEG/FFMpeg.cs [hunk 5]
```diff
@@ -380,38 +382,54 @@ public VideoInfo ReplaceAudio(VideoInfo source, FileInfo audio, FileInfo output,
...
+                _totalTime = TimeSpan.FromSeconds(sources.Sum(source => source.Duration.TotalSeconds));
 
             if (!RunProcess(arguments, output, skipExistsCheck))
                 throw new FFMpegException(FFMpegExceptionType.Conversion, "Could not process file without error");
 
             _totalTime = TimeSpan.MinValue;
+
+            if (output == null)
+                return null;
             return new VideoInfo(output);
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/FFMpeg.cs]: I have already merged a fix for this from WeihanLi.
It is MUCH easier to review a PR if it is focused on one thing, instead of fixing unrelated stuff as well. Those fixes should be separate PRs, so they can be reviewed faster and merged sooner. Now this PR will have to wait until these lines have been reverted, to avoid conflict

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+        public StreamPipeDataReader(System.IO.Stream destanationStream)
+        {
+            DestanationStream = destanationStream;
+        }
+
+        public void ReadData(System.IO.Stream stream)
+        {
+            int read;
+            var buffer = new byte[BlockSize];
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs]: Consider using Stream.CopyTo

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            while ((read = stream.Read(buffer, 0, buffer.Length)) != 0)
+                DestanationStream.Write(buffer, 0, buffer.Length);
+        }
+
+        public async Task ReadDataAsync(System.IO.Stream stream)
+        {
+            int read;
+            var buffer = new byte[BlockSize];
+            while ((read = await stream.ReadAsync(buffer, 0, buffer.Length)) != 0)
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs]: Consider using Stream.CopyTo

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            {
+                pipe.Write(buffer, 0, read);
+            }
+        }
+
+        public async Task WriteDataAsync(System.IO.Stream pipe)
+        {
+            var buffer = new byte[BlockSize];
+            int read;
+            while ((read = await Source.ReadAsync(buffer, 0, buffer.Length)) != 0)
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs]: Consider using Stream.CopyTo

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        public StreamPipeDataWriter(System.IO.Stream stream)
+        {
+            Source = stream;
+        }
+
+        public void WriteData(System.IO.Stream pipe)
+        {
+            var buffer = new byte[BlockSize];
+            int read;
+            while ((read = Source.Read(buffer, 0, buffer.Length)) != 0)
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs]: Consider using Stream.CopyTo

File: FFMpegCore/FFMPEG/Pipes/RawVideoPipeDataWriter.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    /// <summary>
+    /// Implementation of <see cref="IPipeDataWriter"/> for a raw video stream that is gathered from <see cref="IEnumerator{IVideoFrame}"/> 
+    /// </summary>
+    public class RawVideoPipeDataWriter : IPipeDataWriter
+    {
+        public string StreamFormat { get; set; }
+        public int Width { get; set; }
+        public int Height { get; set; }
+        public int FrameRate { get; set; } = 25;
+        private IEnumerator<IVideoFrame> framesEnumerator;
```
Comment (rosenbjerg) [inline FFMpegCore/FFMPEG/Pipes/RawVideoPipeDataWriter.cs:20]: Why IEnumerator instead of IEnumerable?

File: FFMpegCore/FFMPEG/Argument/Atoms/ForceFormatArgument.cs (hunk 2 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Argument/Atoms/ForceFormatArgument.cs:8]: FFMpeg contains much more video codecs and containers then declared in `VideoCodec` enum. And their quantity differs between builds and ffmpeg modules selected for compilation. Users still can use `VideoCodec` enum to instatiate new `ForceFormatArgument` and also use string values. So we can use `ForceFormatArgument("matroska")` and  `ForceFormatArgument(VideoCodec.MpegTs)`.

Definig enums is very user friendly, however you are ruining flexibility of ffmpeg by restricting users to use only predefined sets of codecs.

And what about #49. It is very neccesary argument for a lot of cases but not this. If we force users to use `CustomArgument("-f matroska")` instead of `ForceFormatArgument("matroska")` it would be a kind of workaround or hack around restictions that are provided by only using enum for `ForceFormatArgument`.

More over, we should remember that just a single argument of the same type can by added to `ArgumentContainer` cause it `Dictionary<Type, Argument>` under the hood.

File: FFMpegCore/FFMPEG/Argument/ArgumentContainer.cs (hunk 3 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Argument/ArgumentContainer.cs]: Sure b4dcd9f

File: FFMpegCore/FFMPEG/Argument/Atoms/VideoCodecArgument.cs (hunk 4 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Argument/Atoms/VideoCodecArgument.cs:8]: As i mensioned above it is not the best solution

File: FFMpegCore/FFMPEG/FFMpeg.cs (hunk 5 above)
Comment (max619) [reply FFMpegCore/FFMPEG/FFMpeg.cs]: When i was writiting this code #50 was not merged. But without this line it would be impossible to run code at all

Comment (max619) [reply FFMpegCore/FFMPEG/FFMpeg.cs]: Merged at ee16064

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs (hunk 6 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs]: Fixed at f132a3b

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs (hunk 7 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Pipes/StreamPipeDataReader.cs]: Fixed at f132a3b

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs (hunk 8 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs]: Fixed at f132a3b

File: FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs (hunk 9 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Pipes/StreamPipeDataWriter.cs]: Fixed at f132a3b

File: FFMpegCore/FFMPEG/Pipes/RawVideoPipeDataWriter.cs (hunk 10 above)
Comment (max619) [reply FFMpegCore/FFMPEG/Pipes/RawVideoPipeDataWriter.cs:20]: Because i'm using first frame in `GetFormat()` method and then same frame should be written to the video stream

Comment (rosenbjerg) [reply FFMpegCore/FFMPEG/Pipes/RawVideoPipeDataWriter.cs:20]: Cool. I just wanted to hear if there was a reason for not using the very commin IEnumerable

**Comment by amunim** [comment]
Hi, is there any progress after last review 3 days ago?
Is there any thing I could do to speed it up because I need to encode bitmap images to video also in a very small to size to transfer over the network.
